core/object/thread_runner.cpp

```diff
@@ -0,0 +1,207 @@
+	task->native_func_userdata = p_userdata;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+ThreadRunner::TaskID ThreadRunner::add_task(const Callable &p_action, const String &p_description) {
+	task_mutex.lock();
+	// Get a free task
+	Task *task = task_allocator.alloc();
+	TaskID id = last_task++;
+	task->callable = p_action;
+	task->description = p_description;
+	tasks.insert(id, task);
+	tasks_waiting.add_last(&task->task_elem);
+	task_mutex.unlock();
+
+	task_available_semaphore.post();
+	return id;
+}
+
+bool ThreadRunner::is_task_completed(TaskID p_task_id) const {
+	task_mutex.lock();
+	const Task *const *taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
+		task_mutex.unlock();
+		ERR_FAIL_V_MSG(false, "Invalid Task ID"); // Invalid task
+	}
+
+	bool completed = (*taskp)->completed;
+	task_mutex.unlock();
+
+	return completed;
+}
+
+void ThreadRunner::wait_for_task_completion(TaskID p_task_id) {
+	task_mutex.lock();
+	Task **taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
+		task_mutex.unlock();
+		ERR_FAIL_MSG("Invalid Task ID"); // Invalid task
+	}
+	Task *task = *taskp;
+
+	if (task->waiting) {
+		String description = task->description;
+		task_mutex.unlock();
+		if (description.is_empty()) {
+			ERR_FAIL_MSG("Another thread is waiting on this task: " + itos(p_task_id)); // Invalid task
+		} else {
+			ERR_FAIL_MSG("Another thread is waiting on this task: " + description + " (" + itos(p_task_id) + ")"); // Invalid task
+		}
+	}
+
+	task->waiting = true;
+
+	task_mutex.unlock();
+
+	int *index = thread_ids.getptr(Thread::get_caller_id());
+
+	if (index) {
+		// We are an actual process thread, we must not be blocked so continue processing stuff if available.
+		while (true) {
+			if (task_available_semaphore.try_wait()) {
+				// Solve tasks while they are around.
+				_process_task();
+				continue;
+			}
+			if (task->done_semaphore.try_wait()) {
+				// If done, exit
+				break;
+			}
+			OS::get_singleton()->delay_usec(1); // Microsleep, this could be converted to waiting for multiple objects in supported platforms for a bit more performance.
+		}
+	} else {
+		task->done_semaphore.wait();
+	}
+
+	task_mutex.lock();
+	tasks.erase(p_task_id);
+	task_allocator.free(task);
+	task_mutex.unlock();
+}
+
+void ThreadRunner::init(int p_thread_count) {
+	ERR_FAIL_COND(threads.size() > 0);
+	if (p_thread_count < 0) {
+		p_thread_count = OS::get_singleton()->get_default_thread_pool_size();
+	}
+
+	threads.resize(p_thread_count);
+
+	for (uint32_t i = 0; i < threads.size(); i++) {
+		threads[i].thread.start(&ThreadRunner::_thread_function, &threads[i]);
+	}
+}
+
+void ThreadRunner::finish() {
+	if (threads.size() == 0) {
+		return;
+	}
+
+	task_mutex.lock();
+	SelfList<Task> *E = tasks_waiting.first();
+	while (E) {
+		print_error("Task waiting was never re-claimed: " + E->self()->description);
```

Comment:
This isn't really an error.   It should be possible to submit a bunch of long running tasks and then exit the program.  For that to work, you'd need an additional (optional) Callable in add_task(...) so that there is a separate function to call for cancellation.   At shut down, all pending tasks get called on the cancellation callback.  Waiting on a task that was canceled must then also survive without error, because some of the user code may at the same time call wait_for_task_completion(...) during the shutdown process, based on timing.   If ThreadRunner was RefCounted, you could probably just keep all the task objects that were canceled, with the semaphores posted.   Then these deallocate only when all customers of the ThreadRunner are gone.

Reply:
Long running tasks IMO do not belong here and they should use dedicated system threads. I want to work on a specific interface for this separate to this one, since we have some use cases for it. Of course this should be documented properly.

Reply:
If I read it correctly, all scheduled tasks are in `_tasks_waiting`, even ones that nobody is blocked on.   This could be a lot of stuff, including expensive things like reading files etc.  I think you would want to cancel scheduled work that never started without having to wait for it to actually run?

Reply:
More context in case it is useful:

An async API should always have a guarantee of _exactly_ one callback.   This is so that user code can allocate stuff for the async request, stick it in user data, and clean it up on callback.   When the system shuts down, this state needs to be cleaned up, but it may not be possible to execute the queued work, because the system is going down.  

Additionally, it is just slow to have shut down wait for a bunch of work that nobody is going to use anyway.  That's why there normally is a callback parameter that tells you if you are supposed to just clean up or actually do your work, or you can have an additional callback (Callable) to call if the work item / task gets canceled before running.  Because you can't force GD clients to have a particular callback signature, I think the optional Callable is the better way here.  In that case, the guarantee will be that either your work function gets called or the cancellation callback.  Either way, you know you can clean up whatever related stuff you may have.